### PR TITLE
#11397 fix bug of wrong route result

### LINKED ISSUE #11397: route result is wrong
I execute a SQL like this : "select * from demo_db where num=1 or num=2 and count=3"，
I config 2 dataSources : ds_0 and ds_1
table strategy : tb_${num % 2}，
database strategy : ds_${count % 2}
but route result only contains ds_1，following is actual SQL:
Actual SQL: ds_1 ::: select * from tb_1 where num = 1 or num=2 and count =3
Actual SQL: ds_1 ::: select * from tb_0 where num = 1 or num=2 and count =3
Version : 4.1.0

### PR DESCRIPTION
**Pull request by vince7839**
Fixes #11397.

Changes proposed in this pull request:
- fix bug of wrong route result when an "OR" condition before "AND" confition

### CODE REVIEW

**Review by tristaZero — CHANGES_REQUESTED** [review]
@vince7839 Hi, thanks for your bug-fix. I am glad to see your feedback after you raised the issue. Could you add a UT in `ExpressionBuilderTest` to verify your modification?
@strongduanmu Could you have a review here?

**Comment by strongduanmu** [comment]
@vince7839 Thank you for your pr, can you add some test case(unit test, rewrite test) for it?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/ExpressionBuilder.java [hunk 1]
```diff
@@ -50,13 +52,7 @@ public OrPredicateSegment extractAndPredicates() {
...
                 result.getAndPredicates().addAll(rightBuilder.extractAndPredicates().getAndPredicates());
             } else if (logicalOperator.isPresent() && LogicalOperator.AND == logicalOperator.get()) {
-                ExpressionBuilder leftBuilder = new ExpressionBuilder(((BinaryOperationExpression) expression).getLeft());
-                ExpressionBuilder rightBuilder = new ExpressionBuilder(((BinaryOperationExpression) expression).getRight());
-                for (AndPredicate eachLeft : leftBuilder.extractAndPredicates().getAndPredicates()) {
-                    for (AndPredicate eachRight : rightBuilder.extractAndPredicates().getAndPredicates()) {
-                        result.getAndPredicates().add(createAndPredicate(eachLeft, eachRight));
-                    }
-                }
+                processAndExpression(result);
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/ExpressionBuilder.java:55]: @vince7839 Maybe `result.getAndPredicates().addAll(createAndPredicates())` is better.

Comment (vince7839) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/ExpressionBuilder.java:55]: Do you mean changing the method of "processAndExpression()" to "createAndPredicates()"?

Comment (strongduanmu) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/ExpressionBuilder.java:55]: > Do you mean changing the method of "processAndExpression()" to "createAndPredicates()"?

@vince7839 Yes.

Comment (vince7839) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/ExpressionBuilder.java:55]: Okay,I'll do it and add UT

**Pull request closed without merging**